tasks/fullStack/app/controllers/reviews_controller.rb

```diff
@@ -1,29 +1,45 @@
 class ReviewsController < ApplicationController
+  before_action :set_product, only: [:new]
 
   DEFAULT_TAGS = ['default']
 
   def index
-    if params[:shop_id].present? && Shop.where("id = #{params[:shop_id]}").present?
+    params[:shop_id] ||= 1
+    @shop = Shop.find(params[:shop_id])
+
+    if @shop.present?
       params[:per_page] ||= 10
       offset = params[:page].to_i * params[:per_page]
 
+      @shops = Shop.where.not(id: params[:shop_id]).limit(5)
+
       @data = []
-      products = Product.where("shop_id = #{params[:shop_id]}").sort_by(&:created_at)[offset..(offset + params[:per_page])]
+      products = Product.where("shop_id = #{params[:shop_id]}").limit(5).sort_by(&:created_at)[offset..(offset + params[:per_page])]
       products.each do |product|
-        reviews = product.reviews.sort_by(&:created_at)[offset..(offset + params[:per_page])]
+        reviews = product.reviews.limit(5).sort_by(&:created_at)[offset..(offset + params[:per_page])]
         @data << { product: product, reviews: reviews }
       end
     end
   end
 
+  def new
+    @review = Review.new
+  end
+
   def create
     # TODO: Create reviews in background. No need to show errors (if any) to users, it's fine to skip creating the review silently when some validations fail.
 
-    tags = tags_with_default(params)
-    Review.create(product_id: params[:product_id], body: params[:body], rating: params[:rating], reviewer_name: params[:reviewer_name], tags: tags)
+    tags = tags_with_default(review_params)
+    SubmitReviewJob.perform_async(review_params.to_json)
 
     flash[:notice] = 'Review is being created in background. It might take a moment to show up'
-    redirect_to action: :index, shop_id: Product.find_by(id: params[:product_id]).shop_id
+    redirect_to action: :index, shop_id: Product.find_by(id: review_params[:product_id]).shop_id
+  end
+
+  def average
```

Comment:
nitpick: based on how you implemented it, maybe this endpoint should be moved outside ReviewsController to ShopsController and have the route defined as
```
resources :shops, only: [] do
  member do
    get :average_rating
  end 
end
 ```